packages/server/src/api/segments/segments.service.ts

```diff
@@ -768,12 +722,86 @@ export class SegmentsService {
         this.error(e, this.updateCustomerSegments.name, session);
       }
     }
-    this.debug(
-      `updated all segments with: ${customerId}`,
-      this.updateCustomerSegments.name,
-      session,
-      account.id
-    );
+    return { added: addedToSegments, removed: removedFromSegments };
+  }
+
+  private containsString(obj, searchString) {
+    if (typeof obj === 'string') {
+      return obj.includes(searchString);
+    }
+
+    if (Array.isArray(obj)) {
+      return obj.some((item) => this.containsString(item, searchString));
+    }
+
+    if (typeof obj === 'object' && obj !== null) {
+      return Object.values(obj).some((value) =>
+        this.containsString(value, searchString)
+      );
+    }
+
+    return false;
+  }
+
+  public async updateCustomerSegmentsUsingEvent(
+    account: Account,
+    event: any,
+    customerId: string,
+    session: string,
+    queryRunner: QueryRunner
+  ) {
+    const addedToSegments: Segment[] = [];
+    const removedFromSegments: Segment[] = [];
+    const segments = await this.getSegments(account, undefined, queryRunner);
+
+    for (const segment of segments) {
+      // We skip manual segments and empty inclusion criteria
+      if (segment.type && (segment.type === 'manual' || segment.isUpdating)) {
+        continue;
+      }
+      if (
+        segment.inclusionCriteria &&
+        Object.keys(segment.inclusionCriteria).length === 0
+      ) {
+        continue;
+      }
+      if (!this.containsString(segment.inclusionCriteria, event.event)) {
+        continue;
+      }
+
+      const doInclude = await this.customersService.checkCustomerMatchesQuery(

```

Comment:
This query will run after every event that gets processed, so it might be slow for high throughput of events. We should consider only doing an update if the segment's `inclusionCriteria` is of type event, and skip for attribute-only conditions.